### PR TITLE
[BREAKING] Disallow declarations with name "this", "super" and "_"

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/9604

### CODE REVIEW

File: test/libsolidity/syntaxTests/duplicateFunctions/illegal_names.sol [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+	function this() public {
+	}
+	function super() public {
+	}
+	function _() public {
+	}
+}
+// ----
+// DeclarationError 3726: (14-41): The name "this" is reserved.
+// DeclarationError 3726: (43-71): The name "super" is reserved.
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/duplicateFunctions/illegal_names.sol]: One reason why we allowed `this` and `super` is to be able to define interfaces for contracts written in other languages that do not assign this special meaning to these terms. Maybe we can find a middle ground to at least allow external functions with these names?

File: docs/080-breaking-changes.rst [hunk 2]
```diff
@@ -51,3 +51,5 @@ New Restrictions
 * The global functions ``log0``, ``log1``, ``log2``, ``log3`` and ``log4`` have been removed.
 
   These are low-level functions that were largely unused. Their behaviour can be accessed from inline assembly.
+
+* Declarations with the name ``this``, ``super`` and ``_`` are disallowed.
```
Comment (chriseth) [inline docs/080-breaking-changes.rst]: Do we want to list the exceptions?

**Comment by chriseth** [comment]
Needs rebase.

File: docs/080-breaking-changes.rst (hunk 2 above)
Comment (hrkrshnn) [reply docs/080-breaking-changes.rst]: I added the list of exceptions. Also in changelog.

**Comment by hrkrshnn** [comment]
Rebased and added a test for internal functions with name `super`, `this` and `_`.

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 3]
```diff
@@ -509,6 +509,29 @@ bool DeclarationRegistrationHelper::registerDeclaration(
...
+		}
+		else if (dynamic_cast<EventDefinition const*>(&_declaration))
+		{
+		}
+		else
+			_errorReporter.declarationError(
+				3726_error,
+				*_errorLocation,
+				"The name \"" + name + "\" is reserved."
+			);
```
Comment (bshastry) [inline libsolidity/analysis/NameAndTypeResolver.cpp:536]: I wonder if something like a filtering lambda function that returns true for error will be more readable than empty if-else blocks

Comment (hrkrshnn) [reply libsolidity/analysis/NameAndTypeResolver.cpp:536]: I converted it into a lambda.

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 4]
```diff
@@ -509,6 +509,32 @@ bool DeclarationRegistrationHelper::registerDeclaration(
...
+
+	if (illegalNames.count(name))
+	{
+		auto isPublicFunctionOrEvent = [](Declaration const* _d) -> bool
+		{
+			if (auto functionDefinition = dynamic_cast<FunctionDefinition const*>(_d))
+				if (!functionDefinition->isFree() && functionDefinition->isPublic())
+					return true;
+
+			if (dynamic_cast<EventDefinition const*>(_d))
```
Comment (bshastry) [inline libsolidity/analysis/NameAndTypeResolver.cpp]:
```suggestion
			else if (dynamic_cast<EventDefinition const*>(_d))
```

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 5]
```diff
@@ -509,6 +509,32 @@ bool DeclarationRegistrationHelper::registerDeclaration(
...
+				return true;
+
+			return false;
```
Comment (bshastry) [inline libsolidity/analysis/NameAndTypeResolver.cpp:527]:
```suggestion
			else
				return false;
```

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 3 above)
Comment (bshastry) [reply libsolidity/analysis/NameAndTypeResolver.cpp:536]: Thanks, just a minor suggestion for if-else stmt.

File: libsolidity/analysis/NameAndTypeResolver.cpp (hunk 5 above)
Comment (hrkrshnn) [reply libsolidity/analysis/NameAndTypeResolver.cpp:527]: Fixed the other ones. The problem with this is that when the function is free or internal, this `else` branch won't be visited.

Comment (bshastry) [reply libsolidity/analysis/NameAndTypeResolver.cpp:527]: Ah right. What about something like

```
			if (auto functionDefinition = dynamic_cast<FunctionDefinition const*>(_d); !functionDefinition->isFree() && functionDefinition->isPublic())
				return true;
			else if (dynamic_cast<EventDefinition const*>(_d))
				return true;
			else
				return false;
```

Comment (hrkrshnn) [reply libsolidity/analysis/NameAndTypeResolver.cpp:527]: I feel that the first if statement is less readable.

**Review by bshastry — COMMENTED** [review]
Could you please add tests for the following. Use of illegal names for

- function input/return parameters
- inline assembly e.g., function, variable etc
- enum/immutable
- using declaration e.g., `using super for *;` where `super` is a library

**Comment by hrkrshnn** [comment]
Added tests mentioned by @bshastry

**Comment by bshastry** [comment]
Tests are failing. May just need rebasing

**Comment by hrkrshnn** [comment]
Tests are working now after the rebase.

**Review by chriseth — APPROVED** [review]

**Pull request merged**